File: python/src/aoc/year2015/day9.py
from itertools import permutations
from aoc.util import load_example, load_input


def prepare_data(lines):
    locations = {}
    distances = {}
    for line in lines:
        start, _, dest, _, distance = line.split()
        locations[start] = True
        locations[dest] = True
        distances[(start, dest)] = int(distance)
        distances[(dest, start)] = int(distance)
    return list(locations.keys()), distances


def calc_distance(distances, route):
    return sum(distances[(route[i - 1], route[i])] for i in range(1, len(route)))


def run(lines, fun):
    """
    >>> run(load_example(__file__, '9'), min)
    605
    >>> run(load_example(__file__, '9'), max)
    982
    """
    locations, distances = prepare_data(lines)
    return fun(calc_distance(distances, route) for route in permutations(locations))


def part1(lines):
    return run(lines, min)


def part2(lines):
    return run(lines, max)


if __name__ == "__main__":
    data = load_input(__file__, 2015, "9")
    print(part1(data))
    print(part2(data))

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2015/Day9.kt
package io.github.ocirne.aoc.year2015

import io.github.ocirne.aoc.AocChallenge
import io.github.ocirne.aoc.permutations

typealias Edge = Pair<String, String>

typealias Route = List<String>

class Day9(val lines: List<String>) : AocChallenge(2015, 9) {

    private val locations = HashSet<String>()

    private val distances = HashMap<Edge, Int>()

    init {
        lines.forEach { line ->
            val (start, _, dest, _, distance) = line.split(' ')
            locations.add(start)
            locations.add(dest)
            distances[start to dest] = distance.toInt()
            distances[dest to start] = distance.toInt()
        }
    }

    override fun part1(): Int {
        return run().minOrNull()!!
    }

    override fun part2(): Int {
        return run().maxOrNull()!!
    }

    private fun calcDistance(route: Route): Int {
        return route.zipWithNext { from, to -> distances[from to to]!! }.sum()
    }

    private fun run(): List<Int> {
        return locations.permutations().map { route -> calcDistance(route) }
    }
}
